### ISSUE TITLE
Performance transfer to remote host

### ISSUE THREAD

**Issue by 74cmonty** [jbd/msrsync#9]
Hi,
I'm running 2 hosts with identical HW.
The NIC supports a bandwith of >20Gbits/sec.
Please check the output of iperf below.
```
ld4465:~ # iperf2 -c 192.168.100.11
------------------------------------------------------------
Client connecting to 192.168.100.11, TCP port 5001
TCP window size:  325 KByte (default)
------------------------------------------------------------
[  3] local 192.168.100.13 port 60636 connected with 192.168.100.11 port 5001
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0-10.0 sec  25.1 GBytes  21.5 Gbits/sec
```
I used NFS mount to connect the target server to the source server.
Then I started msrsync with these parameters:
`ld4464:/backup # msrsync -P -p 16 /backup/ML1/ /mnt/ML1/`

When I check the number of rsync-processes on the source host, I get only 9!
```
ld4464:~ # ps -ef | grep "rsync -a"
root     73333 73230 87 14:46 pts/1    00:07:56 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tmp/msrsync-tq0w9Z/tmpesZxD1 /backup/ML1 /mnt/ML1/
root     73334 73232 89 14:46 pts/1    00:08:10 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tmp/msrsync-tq0w9Z/tmpfbXcjJ /backup/ML1 /mnt/ML1/
root     73335 73231 85 14:46 pts/1    00:07:48 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tmp/msrsync-tq0w9Z/tmpvQTsDf /backup/ML1 /mnt/ML1/
root     73339 73334  0 14:46 pts/1    00:00:00 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tmp/msrsync-tq0w9Z/tmpfbXcjJ /backup/ML1 /mnt/ML1/
root     73340 73333  0 14:46 pts/1    00:00:00 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tmp/msrsync-tq0w9Z/tmpesZxD1 /backup/ML1 /mnt/ML1/
root     73341 73339 95 14:46 pts/1    00:08:42 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tmp/msrsync-tq0w9Z/tmpfbXcjJ /backup/ML1 /mnt/ML1/
root     73342 73340 92 14:46 pts/1    00:08:26 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tmp/msrsync-tq0w9Z/tmpesZxD1 /backup/ML1 /mnt/ML1/
root     73343 73335  0 14:46 pts/1    00:00:00 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tmp/msrsync-tq0w9Z/tmpvQTsDf /backup/ML1 /mnt/ML1/
root     73344 73343 92 14:46 pts/1    00:08:26 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tmp/msrsync-tq0w9Z/tmpvQTsDf /backup/ML1 /mnt/ML1/
```

Why are only 9 processes running?

**Comment by jbd** [comment]
Hello, nice setup =)

New rsync (hence new pids) processes are run each time a bucket of file is ready for consumption with no more than `-p` rsync processes running at the same time. Once an rsync process has consumed its files, it finishes and it is not used again. 

You can see less than the requested process number for multiple reasons:

- the filesystem walk has generated just enough work for 9 rsync processes at the time you ran the `ps | grep` command. The filesystem walk could still be running. I guess that if you run the same `ps | grep` command later, you'll see more rsync processes (and new pids). Give it some time, and check again for the number of rsync processes, and look for new pids.

- you check just after, and just before new rsync processes has been spawned. That's unlikely 7 processes have finished at the same time AND you did your check just before new ones have started, but who knows...

To confirm what I just said, you could run something like the following to see the evolution of the number of rsync processes over time:
```
$ while true; do ps -ef|grep "rsync -a" |wc -l; sleep 2; done
```

**Comment by 74cmonty** [comment]
Hi,
I used watch -n 2 'ps -ef | grep "rsync -a"' (I guess this is doing the same as the while.
 ```
Every 2.0s: ps -ef | grep "rsync -a"                                            Thu Jun 28 15:33:43 2018

root     20674 62212  0 15:33 pts/4    00:00:00 watch -n 2 ps -ef | grep "rsync -a"
root     20937 20674  0 15:33 pts/4    00:00:00 watch -n 2 ps -ef | grep "rsync -a"
root     20938 20937  0 15:33 pts/4    00:00:00 sh -c ps -ef | grep "rsync -a"
root     20940 20938  0 15:33 pts/4    00:00:00 grep rsync -a
root     73333 73230 78 14:46 pts/1    00:37:05 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tm
root     73334 73232 79 14:46 pts/1    00:37:28 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tm
root     73335 73231 81 14:46 pts/1    00:38:15 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tm
root     73339 73334  0 14:46 pts/1    00:00:00 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tm
root     73340 73333  0 14:46 pts/1    00:00:00 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tm
root     73341 73339 85 14:46 pts/1    00:40:03 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tm
root     73342 73340 84 14:46 pts/1    00:39:34 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tm
root     73343 73335  0 14:46 pts/1    00:00:00 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tm
root     73344 73343 87 14:46 pts/1    00:41:22 /usr/bin/rsync -a --numeric-ids --from0 --files-from=/tm
```
Could this be related to the number of files on the source to be processed?
```
ld4464:~ # ls -lh /backup/ML1/data/DB_ML1/
total 3.1T
-rw-r----- 1 4029269 sapsys 152K May  5 18:38 AfterDowntime_b4Postproc_databackup_0_1
-rw-r----- 1 4029269 sapsys 1.1T May  5 19:15 AfterDowntime_b4Postproc_databackup_2_1
-rw-r----- 1 4029269 sapsys  81M May  5 18:38 AfterDowntime_b4Postproc_databackup_3_1
-rw-r----- 1 4029269 sapsys  81M May  5 18:38 AfterDowntime_b4Postproc_databackup_4_1
-rw-r----- 1 4029269 sapsys 152K May  8 04:56 After_S4HFPS01_UPD_databackup_0_1
-rw-r----- 1 4029269 sapsys 1.1T May  8 05:30 After_S4HFPS01_UPD_databackup_2_1
-rw-r----- 1 4029269 sapsys  81M May  8 04:56 After_S4HFPS01_UPD_databackup_3_1
-rw-r----- 1 4029269 sapsys  81M May  8 04:56 After_S4HFPS01_UPD_databackup_4_1
-rw-r----- 1 4029269 sapsys 152K Apr 19 09:58 COMPLETE_DATA_BACKUP_databackup_0_1
-rw-r----- 1 4029269 sapsys 1.0T Apr 19 10:35 COMPLETE_DATA_BACKUP_databackup_2_1
-rw-r----- 1 4029269 sapsys  81M Apr 19 09:58 COMPLETE_DATA_BACKUP_databackup_3_1
-rw-r----- 1 4029269 sapsys  81M Apr 19 09:58 COMPLETE_DATA_BACKUP_databackup_4_1

```

**Comment by jbd** [comment]
Are you saying that there is only 12 files to sync in total ?

**Comment by 74cmonty** [comment]
Yes, in this test there are only 12 files.
Checking the target dir only the 3 largest are changing.

**Comment by jbd** [comment]
OK. 

TLDR; `rsync`, hence `msrync` work at the file level, and won't break large file into chunks to transfer them in parallel. 

`msrync` builds buckets of a number of files (1000 by default) OR limit the total size of thoses buckets to 1GB by default. Look at the `-f, --files` and `-s, --size` options: 

```
-f, --files ...       limit buckets to <files> files number [1000]
-s, --size ...        limit partitions to BYTES size (1024 suffixes: K, M, G, T, P, E, Z, Y) [1G]
```

With the source directory you showed, something like this is happening:

- `msrsync` walks the `/backup/ML1/ tree`
- it encounters `AfterDowntime_b4Postproc_databackup_0_1` and `AfterDowntime_b4Postproc_databackup_2_1` files (152K and 1.1T) files. Since it's > 1GB, the bucket is build and consumed by one rsync process
- etc...

In the end, you have 5 buckets created. Something like (could be something else, the order of the filesystem walk is not guaranteed)

```
152K AfterDowntime_b4Postproc_databackup_0_1
1.1T  AfterDowntime_b4Postproc_databackup_2_1
```
```
81M AfterDowntime_b4Postproc_databackup_3_1
81M AfterDowntime_b4Postproc_databackup_4_1
152K After_S4HFPS01_UPD_databackup_0_1
1.1T After_S4HFPS01_UPD_databackup_2_1
```
```
1.1T After_S4HFPS01_UPD_databackup_2_1
```
```
81M After_S4HFPS01_UPD_databackup_3_1
81M After_S4HFPS01_UPD_databackup_4_1
152K COMPLETE_DATA_BACKUP_databackup_0_1
1.0T COMPLETE_DATA_BACKUP_databackup_2_1
```
```
81M COMPLETE_DATA_BACKUP_databackup_3_1
81M COMPLETE_DATA_BACKUP_databackup_4_1
```

In your `ps|grep` output, there is 9 `rsync` processes. But if you look at the full command line, there is only 3 differents commands (look at the `--files-from` parameter). That's because the `rsync` command is forking, it has been designed that way. If you want details, you can have a look at the `rsync(1)` manpage.

`msrsync` is only a wrapper around the `rsync` command. `rsync` by itself (and hence `msrsync`) will not parallelize the transfer of a 1TB file : it will be one copy stream since it treats each file individually. If you want to parallelize the transfer of big files by copying chunks of it in parallel, you should use another tool like `mcp` (https://software.nasa.gov/software/ARC-16494-1). 

`msrsync` advantages resides in its "fire & forget" philosophy and the no dependencies besides `rsync` and `python 2.6`. Under the hood, it run vanilla rsync processes. Hence msrsync benefits from all the goodness of it (maturity/versatility/options) but also the its weaknesses, like moderate performance comparing to a simple `cp` in the case of an initial synchronisation.

**Comment by 74cmonty** [comment]
Thanks for your explanation.
Could I say that msrsync is boosting the mass data transfer the more files I have transfer?
Means it's beneficial when transferring 10000x 1G files vs. 10x 1TB files?

**Comment by jbd** [comment]
It always depends on multiple factor. `msrsync` will really shine when you've got a zillions of small files. But it could also be useful with huge files, only because it can copy them in a parallel fashion without too much thinking: just run `msrsync` with a reasonnable `-p` option (it depends on your source and destination I/O capabilities) and it will give you a significant performance improvement over a single rsync processes, whatever the file size distribution.

The project README performance section gives an example with a duplicated linux source tree: https://github.com/jbd/msrsync/blob/master/README.md#performance

In your specific example, with `-p 16` considering that the source and the destination are not limited by i/o performance, I would say that it would be beneficial since you'll transfer 16 1GB files in parallel at any point in time instead of only 10 1TB transfer. 

With `-p 10`, I would say that the 1TB case would be slighlty more performant because only 10 rsync processes will be necessary to transfer file instead of 1000. But I highly doubt that it would be significant in the end.

**Comment by 74cmonty** [comment]
THX
